gateway/ln-gateway/src/lightning/lnd.rs

```diff
@@ -847,6 +831,93 @@ impl ILnRpcClient for GatewayLndClient {
     }
 }
 
+#[derive(Debug)]
+struct PrivatePaymentResult {
+    preimage: Option<Vec<u8>>,
+    last_failed_reason: Option<String>,
+}
+
+async fn attempt_private_payments(
+    routes: QueryRoutesResponse,
+    invoice: PrunedInvoice,
+    client: &mut LndClient,
+) -> Result<PrivatePaymentResult, LightningRpcError> {
+    let total_amt_msat: i64 =
+        invoice
+            .amount
+            .msats
+            .try_into()
+            .map_err(|error| LightningRpcError::FailedPayment {
+                failure_reason: format!("amount exceeds valid LND amount ranges {error:?}"),
+            })?;
+
+    let mut preimage: Option<Vec<u8>> = None;
+    let mut last_failed_reason: Option<String> = None;
+
+    for mut route in routes.routes {
+        // skip any empty routes
+        if route.hops.is_empty() {
+            continue;
+        }
+
+        // add mpp record to route
+        let last = route.hops.last_mut().expect("must have at least one hop");
+
+        last.mpp_record = Some(MppRecord {
+            payment_addr: invoice.payment_secret.to_vec(),
+            total_amt_msat,
+        });
+
+        trace!("Attempting to send payment to route {route:?}");
+
+        debug!("LND payment request sent for invoice {invoice:?}, waiting for payment status...");
```

Comment:
I think it would be good to log the whole invoice at the `debug` level, but also probably good to log just the `payment_hash` at the info level. It should be easy for gateway operators to find in the logs when payments are attempted